csrc/device_lower/utils.cpp

```diff
@@ -1986,6 +1986,83 @@ std::vector<Expr*> getSyncExprs(AsyncOpType async_type, int64_t keep_stages) {
+getIndexIDs(
+    TensorView* producer,
+    TensorView* consumer,
+    const std::unordered_map<IterDomain*, IterDomain*>* c2p) {
+  // First we find the consumer root IDs that map to the producer
+  std::unordered_map<IterDomain*, IterDomain*> c2p_tmp;
+  if (c2p == nullptr) {
+    c2p_tmp =
+        PairwiseLogicalDomainMap(producer, consumer).mapConsumerToProducer();
+    c2p = &c2p_tmp;
+  }
+  // Track the IDs involved in indexing in both producer and consumer
+  std::unordered_set<IterDomain*> consumer_indexing_ids;
+  std::unordered_set<IterDomain*> producer_indexing_ids;
+  for (IterDomain* id : consumer->getMaybeRootDomain()) {
+    auto it = c2p->find(id);
+    if (it == c2p->end()) {
+      continue;
+    }
+    // These are the immediately mapped consumer root and producer logical
+    // IDs. This is a starting point for our later traversals, which will fill
+    // these sets out.
+    consumer_indexing_ids.insert(it->first);
+    producer_indexing_ids.insert(it->second);
+  }
+
+  // Now traverse from the starting set (which, as noted above is a subset of
+  // either the producer logical or consumer root) to the target which is
+  // either the producer allocation domain or the consumer loop domain. These
+  // are the IDs that will actually affect indexing. Any other IDs can be
+  // skipped.
+  auto traverse = [](std::unordered_set<IterDomain*>& indexing_ids,
+                     const std::vector<IterDomain*>& start_domain,
+                     const std::vector<IterDomain*>& target_domain) {
+    for (auto [expr, dir] : IRBFS::getExprsBetween(
+                                {start_domain.begin(), start_domain.end()},
+                                {target_domain.begin(), target_domain.end()},
+                                /*require_all_to_visited=*/false)
+                                .first) {
+      // If there are any indexing IDs in the inputs, count all outputs as
+      // indexing IDs
+      const auto processExpr = [&indexing_ids](
+                                   const std::vector<Val*>& prev_vals,
+                                   const std::vector<Val*>& next_vals) {
+        if (std::any_of(prev_vals.begin(), prev_vals.end(), [&](Val* prev) {
+              auto* id = dynamic_cast<IterDomain*>(prev);
+              return id && indexing_ids.count(id) != 0;
+            })) {
+          for (Val* v : next_vals) {
+            if (auto* id = dynamic_cast<IterDomain*>(v)) {
+              indexing_ids.insert(id);
+            }
+          }
+        }
+      };
+      if (dir == Direction::Forward) {
+        processExpr(expr->inputs(), expr->outputs());
+      } else if (dir == Direction::Backward) {
+        processExpr(expr->outputs(), expr->inputs());
+      } else {
+        NVF_THROW("Found unexpected direction");
+      }
+    }
+  };
+  traverse(
+      producer_indexing_ids,
+      /*start_domain=*/producer->getLogicalDomain(),
+      /*target_domain=*/producer->getMaybeAllocationDomain());
```

Comment:
I'm not still sure what this function is suppose to return.

For example, if the allocation domain is not set, `getMaybeAllocationDomain` is just equal to the logical domain, so this function would just return the initial `producer_indexing_ids`. Am I following correctly?

Reply:
Yes, if there's no allocation domain then this will not modify `producer_indexing_ids`.

Reply:
So, what can we say from the return value? I think I understand the idea, but I guess the real issue here is the true allocation domain may not be return by `getAllocationDomain()`.

Reply:
> So, what can we say from the return value? I think I understand the idea, but I guess the real issue here is the true allocation domain may not be returned by `getAllocationDomain()`.

`getMaybeAllocationDomain()` should be safe to use shouldn't it? In case there is not allocation domain and we don't modify `producer_indexing_ids`, then it will just hold the IDs from the producer logical domain that are mapped to domains in the consumer.

Reply:
I'm not saying this function would throw an error. It's just I don't understand what the return value would mean. Aren't this function suppose to return indices that are indexed? That isn't true when the true allocation domain is not used.

Reply:
What would the true allocation domain be? Is it not getMaybeAllocationDomain? Assuming we don't later set the allocation domain.

Reply:
It's a mess 😆 

https://github.com/NVIDIA/Fuser/blob/main/csrc/id_model/indexing.cpp#L170

Reply:
Oh boy... Do you think it's safe to use that utility here? Note that the producer indexing ids are not used for predicate elimination but I do check them for inlining. Alternatively I could remove this utility and instead write special cases in predicate elimination and inlining for MmaOp.

Reply:
Oh, I thought this is used to exclude unused producer IDs such as the additional broadcast IDs.

I think we could take the logical domain of the producer, get the path from the loop domain to the logical domain and find which IDs of the loop domain are used. Unused loop IDs should be free to ignore with respect to the producer. Since this is about predicates, I don't think allocation domains should matter.

Also, since this is part of lowering, we can use `TensorIndexer` and its traversal graph, which means (I believe) we can get the true loop domain of the expr.

Reply:
Ah, that makes sense. I'll give it a shot

Reply:
> Oh, I thought this is used to exclude unused producer IDs such as the additional broadcast IDs.

We actually don't directly care about the broadcast IDs in the producer for the purposes of predicate elimination. I just have the loop broadcasts in this test because that lets me schedule the smem tensors.

This is the behavior on TOT: in `ProducerConsumerPairAnalyzer` the `c2p_` mapping is a best-effort replay of the producer as the consumer. When there is a mapping in there for a consumer ID, we only predicate if that dimension is parallelized and possibly oversubscribed _and_ its parallel type doesn't match the producer ID it's mapped to. When there is no mapping, if the dimension is parallelized and maybe oversubscribed, we predicate, but if not then we recurse and process its definition(), predicating splits if they might be non-divisible.

The absence of a mapping in `c2p_` could be because of a difference in scheduling, which is the original intention behind this code I believe. However, now that we have the MmaOp's AxisMapping, it could also be because one of the inputs does not contain any producers for this consumer ID. This is like having an implicit broadcast dimension in the producer so 

In this PR, I'm looking at indexing IDs to try and determine whether a mapping in `c2p_` is missing just because of a scheduling mismatch between producer consumer or if it's because the consumer_id is irrelevant to the producer or if it's only a consumer of unmapped root IDs. So it doesn't really matter whether this is the _actual_ indexing path I think, since we're just trying to identify these unmapped producer ID cases.